src/mongoose_acc.erl

```diff
@@ -1,300 +1,263 @@
+        from_jid => jid:jid(), % optional
+        to_jid => jid:jid() % optional
+       }.
+
+-type strip_params() :: #{
+        lserver := jid:lserver(),
+        element := exml:element(),
+        from_jid => jid:jid(), % optional
+        to_jid => jid:jid() % optional
+       }.
+
+-type stanza_params() :: #{
+        element := exml:element() | undefined,
+        from_jid => jid:jid(), % optional
+        to_jid => jid:jid(), % optional
+        _ => _
+       }.
+
+-type ns_key() :: {NS :: any(), Key :: any()}.
+
+-type strippable() :: boolean() | undefined.
+
+%% --------------------------------------------------------
+%% API
+%% --------------------------------------------------------
+
+-spec new(Params :: new_acc_params()) -> t().
+new(#{ location := Location, lserver := LServer } = Params) ->
+    Stanza = stanza_from_params(Params),
+    #{
+      mongoose_acc => true,
+      ref => make_ref(),
+      timestamp => os:timestamp(),
+      origin_pid => self(),
+      origin_location => Location,
+      origin_stanza => Stanza, % Stanza that triggered the processing
+      stanza => Stanza,
+      lserver => LServer,
+      non_strippable => sets:new()
+     }.
+
+ref(#{ mongoose_acc := true, ref := Ref }) ->
+    Ref.
+
+timestamp(#{ mongoose_acc := true, timestamp := TS }) ->
+    TS.
+
+lserver(#{ mongoose_acc := true, lserver := LServer }) ->
+    LServer.
+
+element(#{ mongoose_acc := true, stanza := #{ element := El } }) ->
+    El;
+element(#{ mongoose_acc := true }) ->
+    undefined.
+
+from_jid(#{ mongoose_acc := true, stanza := #{ from_jid := FromJID } }) ->
+    FromJID;
+from_jid(#{ mongoose_acc := true }) ->
+    undefined.
+
+to_jid(#{ mongoose_acc := true, stanza := #{ to_jid := ToJID } }) ->
+    ToJID;
+to_jid(#{ mongoose_acc := true }) ->
+    undefined.
+
+stanza_name(#{ mongoose_acc := true, stanza := #{ name := Name } }) ->
+    Name;
+stanza_name(#{ mongoose_acc := true }) ->
+    undefined.
+
+stanza_type(#{ mongoose_acc := true, stanza := #{ type := Type } }) ->
+    Type;
+stanza_type(#{ mongoose_acc := true }) ->
+    undefined.
+
+stanza_ref(#{ mongoose_acc := true, stanza := #{ ref := StanzaRef } }) ->
+    StanzaRef;
+stanza_ref(#{ mongoose_acc := true }) ->
+    undefined.
+
+%% While possible, it is not recommended to update stanza to 'undefined'.
+-spec update_stanza(NewStanzaParams :: stanza_params(), Acc :: t()) -> t().
+update_stanza(NewStanzaParams, #{ mongoose_acc := true } = Acc) ->
+    Acc#{ stanza := stanza_from_params(NewStanzaParams) }.
+
+-spec set(Namespace :: any(), K :: any(), V :: any(), Acc :: t()) -> t().
+set(NS, K, V, Acc) ->
+    set(NS, K, V, undefined, Acc).
+ 
+-spec set(Namespace :: any(), K :: any(), V :: any(), IsStrippable :: strippable(), Acc :: t()) ->
+    t().
+set(NS, K, V, undefined, #{ mongoose_acc := true } = Acc) ->
+    Acc#{ {NS, K} => V };
+set(NS, K, V, IsStrippable, #{ mongoose_acc := true, non_strippable := NonStrippable } = Acc) ->
```

Comment:
It's tricky to me what the function does. The name says it's to `set` sth, but in some cases it will remove an element from `non_strippable` set. How many places do we have in the code which calls the function to delete sth from the `non_strippable` set?

Reply:
I think there are none. I can remove this clause.